src/tbb/assert_impl.h

```diff
@@ -37,45 +38,59 @@ namespace tbb {
-    }
-#if _MSC_VER && _DEBUG
-    if (1 == _CrtDbgReport(_CRT_ASSERT, location, line, "tbb_debug.dll", "%s\r\n%s", expression, comment?comment:"")) {
-        _CrtDbgBreak();
-    } else
-#endif
-    {
-        std::fflush(stderr);
-        std::abort();
-    }
-}
 
-// Do not move the definition into the assertion_failure function because it will require "magic statics".
-// It will bring a dependency on C++ runtime on some platforms while assert_impl.h is reused in tbbmalloc 
-// that should not depend on C++ runtime
+// Do not move the definition into the assertion_failure_impl function because it will require "magic statics".
+// It will bring a dependency on C++ runtime on some platforms while assert_impl.h is reused in tbbmalloc
+// that should not depend on C++ runtime. For the same reason, we cannot use std::call_once here.
 static std::atomic<tbb::detail::do_once_state> assertion_state;
 
-void __TBB_EXPORTED_FUNC assertion_failure(const char* location, int line, const char* expression, const char* comment) {
+// TODO: consider extension for formatted error description string
+/* [[noreturn]] */ static void assertion_failure_impl(const char* location, int line, const char* expression, const char* comment) {
 #if __TBB_MSVC_UNREACHABLE_CODE_IGNORED
     // Workaround for erroneous "unreachable code" during assertion throwing using call_once
     #pragma warning (push)
     #pragma warning (disable: 4702)
 #endif
-    // We cannot use std::call_once because it brings a dependency on C++ runtime on some platforms 
-    // while assert_impl.h is reused in tbbmalloc that should not depend on C++ runtime
-    atomic_do_once([&](){ assertion_failure_impl(location, line, expression, comment); }, assertion_state);
+    atomic_do_once([&](){
+        std::fprintf(stderr, "Assertion %s failed (located in the %s function, line in file: %d)\n",
+            expression, location, line);
+
+        if (comment) {
+            std::fprintf(stderr, "Detailed description: %s\n", comment);
+        }
+#if _MSC_VER && _DEBUG
+        if (1 == _CrtDbgReport(_CRT_ASSERT, location, line, "tbb_debug.dll", "%s\r\n%s", expression, comment?comment:"")) {
+            _CrtDbgBreak();
+        } else
+#endif
+        {
+            std::fflush(stderr);
+            std::abort();
+        }
+    }, assertion_state);
 #if __TBB_MSVC_UNREACHABLE_CODE_IGNORED
     #pragma warning (pop)
 #endif
 }
 
+namespace assertion_handler {
+    static std::atomic<assertion_handler_type> handler{assertion_failure_impl}; // initial value is default handler
+
+#if (__TBB_BUILD || __TBBBIND_BUILD) // only TBB and TBBBind use custom handler
+    static assertion_handler_type set(assertion_handler_type new_handler) noexcept {
+        return handler.exchange(new_handler ? new_handler : assertion_failure_impl, std::memory_order_acq_rel);
+    }
+#endif
+
+    static assertion_handler_type get() noexcept {
+        return handler.load(std::memory_order_acquire);
+    }
+}
```

Comment:
As far as I know, we don't add indentation for namespaces:

```suggestion
namespace assertion_handler {
static std::atomic<assertion_handler_type> handler{assertion_failure_impl}; // initial value is default handler

#if (__TBB_BUILD || __TBBBIND_BUILD) // only TBB and TBBBind use custom handler
static assertion_handler_type set(assertion_handler_type new_handler) noexcept {
    return handler.exchange(new_handler ? new_handler : assertion_failure_impl, std::memory_order_acq_rel);
}
#endif

static assertion_handler_type get() noexcept {
    return handler.load(std::memory_order_acquire);
}
} // namespace assertion_handler
```